### ISSUE TITLE
【犀牛鸟实战issue】 Load trained model for inference

### ISSUE THREAD

**Issue by Wi-sc** [Tencent-Hunyuan/Hunyuan3D-2.1#90]
## Issue
【问题描述】使用自己训练的模型进行推理时config与模型不一致
【期望改进】能够使用自己训练的模型进行推理
【完成方式】给出对应代码
【参考链接】https://github.com/Tencent-Hunyuan/Hunyuan3D-2.1/issues/59

## 备注
本issue为2025犀牛鸟开源人才培养活动专属issue，仅供参与犀牛鸟的在校大学生及开发者参与领取，认领前请先报名2025犀牛鸟开源人才培养计划（https://opensource.tencent.com/summer-of-code）。
【认领issue】在issue所在repo页面留言“已成功领取本issue”→微信私信犀牛鸟小助手（活动群-黄贵龙）：issue链接+认领截图→犀牛鸟小助手回复“已成功领取本issue”，即可开始做issue任务。
【完成issue】已完成issue的同学，请同步犀牛鸟小助手（私信活动群-黄贵龙回复“已完成本issue任务+代码合入截图等证明”）。
注意：如7天内无提交任何进展包括不限于comment \ commit \ Pull Request，则将视为同学主动放弃issue，组委会将释放issue给下一位等候者。

**Comment by LIUYellowBlack** [comment]
已成功领取本issue

**Comment by s572915912** [comment]
已成功领取本issue

**Comment by ANGJustinl** [comment]
已成功领取本issue

**Comment by Landialos0122** [comment]
已成功领取本issue

**Comment by LIUYellowBlack** [comment]
已成功领取本issue

**Comment by zjc178** [comment]
已成功领取本issue

**Comment by LIUYellowBlack** [comment]
（因[https://github.com/Tencent-Hunyuan/Hunyuan3D-2.1/blob/main/hy3dshape/configs/hunyuan3ddit-mini-overfitting-flowmatching-dinog518-bf16-lr1e4-512.yaml打不开，所以按照https://github.com/Tencent-Hunyuan/Hunyuan3D-2.1/blob/main/hy3dshape/configs/hunyuandit-mini-overfitting-flowmatching-dinol518-bf16-lr1e4-4096.yaml内容，进行替换）](https://github.com/Tencent-Hunyuan/Hunyuan3D-2.1/blob/main/hy3dshape/configs/hunyuan3ddit-mini-overfitting-flowmatching-dinog518-bf16-lr1e4-512.yaml%E6%89%93%E4%B8%8D%E5%BC%80%EF%BC%8C%E6%89%80%E4%BB%A5%E6%8C%89%E7%85%A7https://github.com/Tencent-Hunyuan/Hunyuan3D-2.1/blob/main/hy3dshape/configs/hunyuandit-mini-overfitting-flowmatching-dinol518-bf16-lr1e4-4096.yaml%E5%86%85%E5%AE%B9%EF%BC%8C%E8%BF%9B%E8%A1%8C%E6%9B%BF%E6%8D%A2%EF%BC%89)
模型框架：
开源项目 [Tencent‑Hunyuan/Hunyuan3D‑2.1](https://github.com/Tencent-Hunyuan/Hunyuan3D-2.1)
提供了基于 Flow Matching 的 3D 重建流水线Hunyuan3DDiTFlowMatchingPipeline，配合官方预训练模型使用。
任务需求：
希望用自己训练得到的 .ckpt 检查点（checkpoint）文件，配合相应的 YAML 配置文件，在推理阶段直接载入并生成 3D 网格。
官方给出的示例演示了如何加载他们预先整理好的预训练模型结构与权重，而我的 .ckpt 与示例里的配置结构不完全一致，导致直接调用 from_single_file 会报错。
代码研究
我的代码[https://huggingface.co/tencent/Hunyuan3D-2.1/blob/main/hunyuan3d-dit-v2-1/config.yaml中将「Denoiser」「VAE」「Conditioner」「Scheduler」「ImageProcessor」「Pipeline」拆成了六个字段，训练时配置把「整个](https://huggingface.co/tencent/Hunyuan3D-2.1/blob/main/hunyuan3d-dit-v2-1/config.yaml%E4%B8%AD%E5%B0%86%E3%80%8CDenoiser%E3%80%8D%E3%80%8CVAE%E3%80%8D%E3%80%8CConditioner%E3%80%8D%E3%80%8CScheduler%E3%80%8D%E3%80%8CImageProcessor%E3%80%8D%E3%80%8CPipeline%E3%80%8D%E6%8B%86%E6%88%90%E4%BA%86%E5%85%AD%E4%B8%AA%E5%AD%97%E6%AE%B5%EF%BC%8C%E8%AE%AD%E7%BB%83%E6%97%B6%E9%85%8D%E7%BD%AE%E6%8A%8A%E3%80%8C%E6%95%B4%E4%B8%AA) Diffuser」当作一个 model，其内部再用 params.denoiser_cfg、params.first_stage_config、params.cond_stage_config、params.scheduler_cfg、params.pipeline_cfg、params.image_processor_cfg 等子字段分装所有组件。
官方的预训练代码：训练时配置把「整个 Diffuser」当作一个 model，其内部再用 params.denoiser_cfg、params.first_stage_config、params.cond_stage_config、params.scheduler_cfg、params.pipeline_cfg、params.image_processor_cfg 等子字段分装所有组件
问题分析
1.官方预训练模型的config.yaml中，model、vae、scheduler、image_processor、conditioner等字段命名和层级与我训练时用的YAML有所差异。
2. from_single_file方法内部直接用OmegaConf读config，config期望的字段路径与我的训练时配置不匹配，于是失败
如何解决
建立关系一一对应：
将我的字段跟预训练模型的字段匹配：
config.model->config.model.params.denoiser_cfg
config.vae->config.model.params.first_stage_config
config.conditioner->config.model.params.cond_stage_config
config.scheduler->config.model.params.scheduler_cfg
config.image_processor->config.model.params.image_processor_cfg
config.pipeline->config.model.params.pipeline_cfg
按照以上逻辑，把训练时的 YAML和checkpoint映射到预训练模型所需的各子模块就能顺利加载并推理
我的代码
import os
import torch
from omegaconf import OmegaConf

ROOT_DIR = r"F:\Code reproduction\Hunyuan3D-2.1"
sys.path.insert(0, ROOT_DIR)
LOCAL_CONFIG = os.path.join(
ROOT_DIR,
"hy3dshape",
"configs",
"hunyuandit-mini-overfitting-flowmatching-dinol518-bf16-lr1e4-4096.yaml",
)
CKPT_PATH = os.path.join(
ROOT_DIR,
"hy3dshape",
"output_folder",
"dit",
"overfitting",
"ckpt",
"ckpt-step=00084000.ckpt",
)

可选：Hugging Face 上的通用配置（如需加载预训练模型）
HF_CONFIG = os.path.join(
ROOT_DIR,
"hunyuan3d-dit-v2-1",
"config.yaml",
)

from hy3dshape.utils import instantiate_from_config
from hy3dshape.pipelines import Hunyuan3DDiTFlowMatchingPipeline

def load_hunyuan3d_pipeline(
ckpt_path: str,
config_path: str,
device: torch.device = torch.device('cuda'),
dtype: torch.dtype = torch.float16,
) -> Hunyuan3DDiTFlowMatchingPipeline:
"""
加载自训练模型进行推理的完整函数。
"""

config = OmegaConf.load(config_path)

ckpt = torch.load(ckpt_path, map_location='cpu')
state_dict = ckpt.get('state_dict', ckpt)

denoiser = instantiate_from_config(config.model.params.denoiser_cfg)
vae = instantiate_from_config(config.model.params.first_stage_config)
image_processor = instantiate_from_config(config.model.params.image_processor_cfg)
conditioner = instantiate_from_config(config.model.params.cond_stage_config)
scheduler_cfg = {
    "target": "hy3dshape.schedulers.FlowMatchEulerDiscreteScheduler",
    "params": {"num_train_timesteps": config.model.params.scheduler_cfg.num_train_timesteps}
}
scheduler = instantiate_from_config(scheduler_cfg)

load_map = [
    (vae, "first_stage_model"),
    (conditioner, "cond_stage_model"),
    (denoiser, "model"),
]
for module, prefix in load_map:
    sub_dict = {
        k.replace(f"{prefix}.", ""): v
        for k, v in state_dict.items()
        if k.startswith(prefix + ".")
    }
    module.load_state_dict(sub_dict, strict=True)

pipeline = Hunyuan3DDiTFlowMatchingPipeline(
    model=denoiser,
    vae=vae,
    scheduler=scheduler,
    conditioner=conditioner,
    image_processor=image_processor,
    device=device,
    dtype=dtype,
)
pipeline.to(device)
pipeline.eval()
return pipeline
if name == 'main':
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
pipeline = load_hunyuan3d_pipeline(
ckpt_path=CKPT_PATH,
config_path=LOCAL_CONFIG,
device=device,
dtype=torch.float16,
)
input_image = os.path.join(ROOT_DIR, 'image.jpg') # 替换为你的输入图像路径
result = pipeline(image=input_image)[0]
output_path = os.path.join(ROOT_DIR, 'output.obj')
result.export(output_path)
print(f"推理完成，结果已保存到: {output_path}")

**Comment by ComistryMo** [comment]
已成功领取本issue

**Comment by Qi-qi0317** [comment]
已成功领取本issue

**Comment by dyw-vince** [comment]
已成功领取本issue

**Comment by WJXTECH** [comment]
已成功领取本issue

**Comment by 1530374515** [comment]
已成功领取本issue

**Comment by LingmaFuture** [comment]
已成功领取本issue

**Comment by Rose-upForever** [comment]
已成功领取本issue

**Comment by jucs8888** [comment]
已成功领取本issue
